Install Parity

croaky opened the pull request:
* The Parity Homebrew [package] installs Heroku Toolbelt, Git, Postgres
  but it seems nice for documentation purposes to
  keep the Homebrew packages installed for the other dependencies.
* Tap the Homebrew formula first, so it updates with `brew update`.

[package]: https://github.com/thoughtbot/homebrew-formulae/blob/master/Formula/parity.rb

croaky commented:
@thoughtbot/parity @thoughtbot/shell your thoughts?

calebthompson commented:
Several questions and comments:

I've mostly given up trying to install postgres with homebrew. I've been using Postgres.app, which I install with homebrew-cask. I think this is common, as several people recommended it to me when last I tried to use laptop.

Were you able to get around the issues of not being able to update these individually?

While I understand its use case, I don't actually use Parity. I don't recall it coming up in recent (a year or so) pairing sessions either, but that could just be not having done much with heroku. Is this something that should be included in laptop? Was there a research card for this? https://trello.com/search?q=parity yields nothing.

calebthompson commented:
> Group Homebrew packages by topic, with comments, for readability.

This is a cool change.

croaky commented:
Thanks for the comments, @calebthompson.

> I've mostly given up trying to install postgres with homebrew. I've been using Postgres.app, which I install with homebrew-cask. I think this is common, as several people recommended it to me when last I tried to use laptop.

Gotcha. I haven't had issues with Postgres through Homebrew that I can remember. In the last year, we've had three Postgres-related changes in the Laptop script that I see:

* oldest: [brew services restart postgresql](https://github.com/thoughtbot/laptop/commit/4b778a2da213997731df01811f73094e4dfe8140)
* old: [use launchtrl instead of no longer supported brew services](https://github.com/thoughtbot/laptop/commit/15355e577a9c502594799e0e4905290d0b4a0c32)
* latest: [drop --no-python flag so we get bottled version of Postgres](https://github.com/thoughtbot/laptop/commit/b16220326396fd9f7df3cf92b000a3a53dd88b96)

For my habits, I've tried homebrew-cask formulas in the past and ended up first restricting their use to non-developer facing programs (Rdio, Bartender, Chrome, etc.), then dropped them as many of them self-update (Chrome), and found that the extra layer of the cask wasn't necessary, for Postgres installs nicely through plain Homebrew.

Sticking to plain Homebrew in the Laptop script also lets us take advantage of some of the shell functions in the script, such as [brew_install_or_upgrade](https://github.com/thoughtbot/laptop/blob/master/mac#L56-L68), which provides nice output.

The only other dotfiles/laptop thing I do with Postgres is have a [restart-postgres](https://github.com/croaky/dotfiles/blob/master/aliases.local#L15) alias in case things get stuck.

> Were you able to get around the issues of not being able to update these individually?

Are you referring to my comment on [this prior pull request](https://github.com/thoughtbot/laptop/pull/390)?

> I discovered an issue with this which is that by making this change, we lose the ability to independently upgrade Git, Postgres, Heroku Toolbelt. Closing and will open a separate smaller change from this branch.

Yup, I closed that one and opened this one to fix that issue.

> While I understand its use case, I don't actually use Parity. I don't recall it coming up in recent (a year or so) pairing sessions either, but that could just be not having done much with heroku. Is this something that should be included in laptop?

The history of Parity in Laptop is:

* [we added it in July last year](https://github.com/thoughtbot/laptop/commit/094e6e7f9e7141bc46271108454dd42a57e00ffe)
* [we removed it in December](https://github.com/thoughtbot/laptop/commit/545d207ba57c953096f23c19f3474313bb6ab7ae)

A few things have changed since December:

* @JoelQ converted Parity into a binary using Traveling Ruby in the [0.5.0 release](https://github.com/thoughtbot/parity/releases/tag/v0.5.0) . We were previously not happy with Laptop installing any gems beyond Bundler.
* We've been maintaining a Parity Homebrew formula in [thoughtbot/formulae](https://github.com/thoughtbot/homebrew-formulae)
* There's a @thoughtbot/parity GitHub group
* A bunch of thoughtbot folks said in Slack that they use Parity frequently and therefore we could "graduate" it from croaky/parity to an official thoughtbot project at [thoughtbot/parity](https://github.com/thoughtbot/parity)

The project was starting to feel like it had crossed an inflection point within the company so I figured I'd test the waters again to see if it made sense to bring it back into Laptop.

Happy to be conservative about it, though, to keep Laptop lean, if that's the stronger opinion of the group.

One other interesting question might be: should we install all [thoughtbot/formulae](https://github.com/thoughtbot/homebrew-formulae) in Laptop? We currently install RCM but not gitsh, parity, liftoff, or pick.

> Was there a research card for this? https://trello.com/search?q=parity yields nothing.

For whether Parity should be in Laptop? Nope. I was thinking this PR would act as that discussion. Hopefully this comment serves to tie together past discussions.

calebthompson commented:
Wow, thank you for your thoughtful response.



> I haven't had issues with Postgres through Homebrew that I can remember

That wasn't really the point of this PR, so let's shelve that discussion. If anything, it should have been a separate PR, sorry.



> For my habits, I've tried homebrew-cask formulas in the past and ended up first restricting their use to non-developer facing programs (Rdio, Bartender, Chrome, etc.), then dropped them as many of them self-update (Chrome), and found that the extra layer of the cask wasn't necessary, for Postgres installs nicely through plain Homebrew.

Interesting. I just started a fresh computer recently and ended up using homebrew + homebrew cask from almost everything I wanted to install. So far it's been nice, mostly as a way to make sure I've got all the apps I wanted installed.

It would be nice to have a homebrew-like interface for installing, but not managing, programs that do their own updating (Chrome, gh/hub, heroku-toolkit). 



> Yup, I closed that one and opened this one to fix that issue.

:+1: 

> The project was starting to feel like it had crossed an inflection point within the company so I figured I'd test the waters again to see if it made sense to bring it back into Laptop.

Sweet, makes sense. I'm all for adding this if the majority of people are using it. I can either remove it or ignore it as I like. Maybe I'll even start accidentally using it ;)



> Happy to be conservative about it, though, to keep Laptop lean, if that's the stronger opinion of the group.

That's how I'd love to see us approach suspenders and laptop, but the prevailing opinion seems to be against me and that's fine.



> > Was there a research card for this? https://trello.com/search?q=parity yields nothing.
>
> For whether Parity should be in Laptop? Nope. I was thinking this PR would act as that discussion. Hopefully this comment serves to tie together past discussions.

No, for whether we should recommend its use generally. I've found that a PR to suspenders, laptop, or dotfiles is often the result of a research card having been concluded as a success, so I wondered if that was something I could find some of the answers to my questions in.

gabebw commented:
I don't use RCM, but I use all of the rest of it. I use parity every single day. I'm :+1: on this, especially grouping packages by category.

bernerdschaefer commented:
:+1: to parity by default!

mike-burns commented:
Looks good to me. Welcome aboard, Parity!

croaky commented:
> > Group Homebrew packages by topic, with comments, for readability.

> This is a cool change.

Extracted that to its own change in https://github.com/thoughtbot/laptop/pull/433

croaky commented:
> Use the --repair flag to check for dead symlinks and relink all valid formulae across taps.

Extracted that to its own change in https://github.com/thoughtbot/laptop/pull/434

gabebw commented on mac:
```diff
@@ -128,34 +128,58 @@ else
 brew update
 
+# Dotfiles
+if ! command -v rcup >/dev/null; then
```
Why wrap the rcm installation in this `if`, but not any of the others?

gabebw replied:
That is -- why is `rcm` special and deserving of this check?

croaky replied:
It's either that `rcm` was added to Laptop before the `brew_install_or_upgrade` function was added or that there's something about `thoughtbot/formulae` that doesn't make them respond correctly to the `brew_is_installed` and `brew_is_upgradable` functions:

```
Installing parity ...
Warning: thoughtbot/formulae/parity-0.9.2.1 already installed

Installing rcm ...
Warning: thoughtbot/formulae/rcm-1.3.0 already installed
```
